### DISCUSSION TITLE
Feature Request: `list_to_struct()`

### DISCUSSION THREAD

**Discussion by cmdlineluser** [duckdb/duckdb#11406] [Ideas]
This has sort of come up a couple of times:

- https://github.com/duckdb/duckdb/discussions/11337
- https://github.com/duckdb/duckdb/discussions/7716

```python
df = duckdb.sql("""
from values (1, [2, 3], 4), (5, [6, 7, 8], 9) t(A, B, C)
""").pl()

# shape: (2, 3)
# ┌─────┬───────────┬─────┐
# │ A   ┆ B         ┆ C   │
# │ --- ┆ ---       ┆ --- │
# │ i32 ┆ list[i32] ┆ i32 │
# ╞═════╪═══════════╪═════╡
# │ 1   ┆ [2, 3]    ┆ 4   │
# │ 5   ┆ [6, 7, 8] ┆ 9   │
# └─────┴───────────┴─────┘
```

Polars for example has a [`list.to_struct()`](https://docs.pola.rs/docs/python/dev/reference/expressions/api/polars.Expr.list.to_struct.html#polars.Expr.list.to_struct) function:

```python
df.with_columns(pl.col("B").list.to_struct(n_field_strategy="max_width"))
# shape: (2, 3)
# ┌─────┬────────────┬─────┐
# │ A   ┆ B          ┆ C   │
# │ --- ┆ ---        ┆ --- │
# │ i32 ┆ struct[3]  ┆ i32 │
# ╞═════╪════════════╪═════╡
# │ 1   ┆ {2,3,null} ┆ 4   │
# │ 5   ┆ {6,7,8}    ┆ 9   │
# └─────┴────────────┴─────┘
```

Which we can UNNEST:

```python
df_struct = df.with_columns(
   pl.col("B").list.to_struct(
      n_field_strategy = "max_width",
      fields = lambda idx: f"B{idx + 1}" # rename fields
   )
)

duckdb.sql("from df_struct select * replace unnest(B) as B")
# ┌───────┬───────┬───────┬───────┬───────┐
# │   A   │  B1   │  B2   │  B3   │   C   │
# │ int32 │ int32 │ int32 │ int32 │ int32 │
# ├───────┼───────┼───────┼───────┼───────┤
# │     1 │     2 │     3 │  NULL │     4 │
# │     5 │     6 │     7 │     8 │     9 │
# └───────┴───────┴───────┴───────┴───────┘
```

Would it be feasible/useful to simplify this transformation in DuckDB?

**Comment by Maxxen** [comment]
This would only really work if you had a fixed-size list right? Otherwise, how would you know how many columns to return without having to scan through all the rows first? DuckDB has to know all the types and schema of a query _before_ it starts execution.

**Reply by cmdlineluser** [reply]
Yeah, I guess `ARRAY -> STRUCT` would make more sense in that regard as it is fixed width.

I thought maybe we could cast but it seems one has to do the length padding manually:

```python
duckdb.sql("from df select * replace B::int[3] as B")
# ConversionException: Conversion Error: Cannot cast list with length 2 to array with length 3
```

Perhaps I'm just thinking about the problem the wrong way and it should be done as a PIVOT or similar:

```python
duckdb.sql("""
pivot (
   from df 
   select 
      B.generate_subscripts(1) as name,
      B[B.generate_subscripts(1)] as value,
      *
)   
on 'B' || name
using first(value)
""")

# ┌───────┬───────────┬───────┬───────┬───────┬───────┐
# │   A   │     B     │   C   │  B1   │  B2   │  B3   │
# │ int32 │  int32[]  │ int32 │ int32 │ int32 │ int32 │
# ├───────┼───────────┼───────┼───────┼───────┼───────┤
# │     5 │ [6, 7, 8] │     9 │     6 │     7 │     8 │
# │     1 │ [2, 3]    │     4 │     2 │     3 │  NULL │
# └───────┴───────────┴───────┴───────┴───────┴───────┘
```

**Reply by soerenwolfers** [reply]
I don't think you're thinking about the problem in the wrong way at all. Just because it is already possible doesn't mean it cannot be made simpler. Just look at the complexity of that pivot. And if you actually want a struct rather than multiple columns, you're not even done yet.
